File: leetcode/go/2-add-two-numbers.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var dummy = new(ListNode)
	var curr = dummy
	var carry int
	var x, y int
	for l1 != nil || l2 != nil {
		if l1 != nil {
			x = l1.Val
			l1 = l1.Next
		} else {
			x = 0
		}
		if l2 != nil {
			y = l2.Val
			l2 = l2.Next
		} else {
			y = 0
		}
		sum := x + y + carry
		carry = sum / 10
		curr.Next = new(ListNode)
		curr = curr.Next
		curr.Val = sum % 10
	}
	if carry > 0 {
		curr.Next = new(ListNode)
		curr.Next.Val = carry
	}
	return dummy.Next
}

File: leetcode/js/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let dummyHead = new ListNode();
    let curr = dummyHead;
    let carry = 0;
    while(l1 || l2) {
        let x = (l1 ? l1.val : 0);
        let y = (l2 ? l2.val : 0);
        sum = x + y + carry;
        curr.next = new ListNode(sum%10);
        carry = Math.floor(sum/10);
        l1 && (l1 = l1.next);
        l2 && (l2 = l2.next);
        curr = curr.next;
    }
    (carry > 0) && (curr.next = new ListNode(carry));
    return dummyHead.next;
};